### PR TITLE
kak-tree-sitter: init at 1.1.2

### PR DESCRIPTION
**Pull request by lelgenio**
## Description of changes

Kakoune extension that adds support for tree-sitter grammars, instead of the basic regex-based default.

Home: https://git.sr.ht/~hadronized/kak-tree-sitter

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by lelgenio** [comment]
Result of `nixpkgs-review pr 333951` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kak-tree-sitter</li>
  </ul>
</details>

File: pkgs/by-name/da/dart-sass/package.nix [hunk 1]
```diff
@@ -85,6 +85,7 @@ buildDartApplication rec {
     description = "Reference implementation of Sass, written in Dart";
     mainProgram = "sass";
     license = lib.licenses.mit;
+    platforms = lib.platforms.unix;
```
Comment (AndersonTorres) [inline pkgs/by-name/da/dart-sass/package.nix:88]: Split this mod in another PR or commit

File: pkgs/by-name/ka/kak-tree-sitter/package.nix [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    repo = "kak-tree-sitter";
+    rev = "kak-tree-sitter-v${version}";
+    hash = "sha256-wBWfSyR8LGtug/mCD0bJ4lbdN3trIA/03AnCxZoEOSA=";
+  };
+
+  cargoHash = "sha256-OQPUWqJAts8DbFNSsC/CmMCbuZ9TVxRTR05O7oiodKI=";
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  # Tree-Sitter grammars are C programs that need to be compiled
```
Comment (AndersonTorres) [inline pkgs/by-name/ka/kak-tree-sitter/package.nix:27]: I know few to nothing about tree-sitter and lsp servers.
How does it work? 
Are C source files compiled when the editor starts, or something like that? Why can't these files be precompiled?

Comment (lelgenio) [reply pkgs/by-name/ka/kak-tree-sitter/package.nix:27]: > How does it work?

Tree sitter grammars are written using a javascript library that generates efficient C code. 
That C code is normally committed in the same repository as the original javascript code.
This weird process makes it easy for developers to write new grammars, and easy for code editors to install new grammars, all you need is a C compiler:
https://github.com/helix-editor/helix/blob/ff33b07756548935577aefc15cf48a1beb27b162/languages.toml#L512-L514

> Are C source files compiled when the editor starts, or something like that?

Something like that, but only once of course, for kak-tree-sitter you can install all declared grammars using `ktsctl sync --all`. If you add another grammar to your config, you can run it again.

> Why can't these files be precompiled?

They certainly can! Many are already packaged in `nixpkgs` [here](https://github.com/NixOS/nixpkgs/tree/master/pkgs/development/tools/parsing/tree-sitter/grammars).

I'm already using them, but it's janky for now: https://git.lelgenio.com/lelgenio/nixos-config/src/commit/9da25c99fe2f4e2255c52f6cdb65b2b54f98e119/user/kakoune/kak-tree-sitter.nix

The problem is that you'll frequently find yourself needing to install some grammar that is not packaged in `nixpkgs` or just don't want nix consume your life :laughing:

Adding `cc` to the `PATH` of `ktsctl` was not mandatory for this package to be useful, but since the consensus on `nixpkgs` is to add optional dependencies by default, I went with that.

Comment (AndersonTorres) [reply pkgs/by-name/ka/kak-tree-sitter/package.nix:27]: Hum, I am a bit skeptical about bringing a C compiler to the scope of a package.
To me it looks more natural to let this decision to the user.

Thinking in the perspective of a user of another distro like Ubuntu or Rhino, theey would install two packages: the C compiler and the kak-tree-sitter.
Maybe a smarter user would create a metapackage for declarative package installation - https://ayats.org/blog/declarative-ubuntu - but this is a whole another topic.

Point is, a typical user is expected to install both packages (plus some book-keeping like the wrapper you did).

Now, back to Nixpkgs.

Another problem is that we are creating an artificial dependency link here: ~if~ when tcc is updated, kak-tree-sitter will be needlessly recompiled (given that kak-tree-sitter does not use tcc as a real dependency upstream).

That being said, a better solution would be creating a third package, say `kak-tree-sitter-plus-tcc`, basically a wrapper that merges both packages in a `symlinkJoin`.
I did something like this in x16-emulator, remember me later.

File: pkgs/by-name/ka/kak-tree-sitter-unwrapped/package.nix [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  lib,
+  rustPlatform,
+  fetchFromSourcehut,
+  nix-update-script,
+  testers,
+  kak-tree-sitter-unwrapped,
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "kak-tree-sitter-unwrapped";
```
Comment (AndersonTorres) [inline pkgs/by-name/ka/kak-tree-sitter-unwrapped/package.nix:11]: This is a matter of taste, but
What about making this the default `kak-tree-sitter` while the wrapped version is called `kak-tree-sitter-wrapped`?

File: pkgs/by-name/ka/kak-tree-sitter-unwrapped/package.nix [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    rev = "kak-tree-sitter-v${version}";
+    hash = "sha256-wBWfSyR8LGtug/mCD0bJ4lbdN3trIA/03AnCxZoEOSA=";
+  };
+
+  cargoHash = "sha256-v0DNcWPoHdquOlyPoPLoFulz66yCPR1W1Z3uuTjli5k=";
+
+  passthru = {
+    updateScript = nix-update-script { };
+    tests.version = testers.testVersion { package = kak-tree-sitter-unwrapped; };
+  };
```
Comment (AndersonTorres) [inline pkgs/by-name/ka/kak-tree-sitter-unwrapped/package.nix:26]: Another idea:
```suggestion
    wrapper = callPackage ./wrapper.nix { };
  };
```

in which the wrapper has the same code you implemented as kak-tree-sitter-wrapped.

This way all files related to kak-tree-sitter are in the same directory.

Finally you can create an alias in all-packages.nix:

```nix
kak-tree-sitter-wrapped = kak-tree-sitter.wrapper;
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ka/kak-tree-sitter-unwrapped/package.nix (hunk 3 above)
Comment (lelgenio) [reply pkgs/by-name/ka/kak-tree-sitter-unwrapped/package.nix:11]: > What about making this the default `kak-tree-sitter` while the wrapped version is called `kak-tree-sitter-wrapped`?

I would not like that, to be honest. My rationale here is that the package should have maximum functionality by default. You install it and you are done. 

If you really cannot afford the 16Mib necessary for `tcc` you can always get rid of it, but that does not justify imposing a worse out-of-the-box experience for other users.

Comment (AndersonTorres) [reply pkgs/by-name/ka/kak-tree-sitter-unwrapped/package.nix:11]: On the other hand the unwrapped version is more composable.
What if the programming environment I am using provides a C compiler upfront?
In this scenario tcc would be redundant.

But as I said it is a matter of taste, since both derivations are still available.

File: pkgs/by-name/ka/kak-tree-sitter-unwrapped/package.nix (hunk 4 above)
Comment (AndersonTorres) [reply pkgs/by-name/ka/kak-tree-sitter-unwrapped/package.nix:26]: On the other hand, I think it is still interesting to write the wrapper in the passthru.

**Review by AndersonTorres — APPROVED** [review]
Result of `nixpkgs-review pr 333951` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>kak-tree-sitter</li>
    <li>kak-tree-sitter-unwrapped</li>
  </ul>
</details>

Final hint: add a test version.
